dsf-gdb/org.eclipse.cdt.tests.dsf.gdb/src/org/eclipse/cdt/tests/dsf/gdb/tests/nonstop/MultiProcessRemoteTest.java

```diff
@@ -0,0 +1,239 @@
+		assertTrue("Process-1 is not attached properly as no threads found for it", ctx != null && ctx.length > 0);
+
+		process = launchApplication();
+		ctx = attachToProcess(Long.toString(process.pid()));
+		assertTrue("Process-2 is not attached properly as no threads found for it", ctx != null && ctx.length > 0);
+
+		process = launchApplication();
+		ctx = attachToProcess(Long.toString(process.pid()));
+		assertTrue("Process-3 is not attached properly as no threads found for it", ctx != null && ctx.length > 0);
+	}
+
+	@Test
+	public void launchGdbServerWithoutProgramAndConnectMultipleExecutables_noProgram() throws Throwable {
+		launchGdbServerWithProgramAndConnectMultipleExecutables();
+	}
+
+	@Test
+	public void launchGdbServerWithProgram_DebugNewExecutable_And_ConnectExecutable() throws Throwable {
+		assertEquals("Wrong number of containers at the start of launch", SyncUtil.getAllContainerContexts().length,
+				getInitialProcessCount());
+
+		final IPath execPath = fGDBBackend.getProgramPath();
+		Map<String, Object> attributes = getLaunchConfiguration().getAttributes();
+
+		IMIExecutionDMContext[] ctx = debugNewProcess(execPath, attributes);
+		assertTrue("Process-1 is not started properly as no threads found for it", ctx != null && ctx.length > 0);
+
+		Process process = launchApplication();
+		ctx = attachToProcess(Long.toString(process.pid()));
+		assertTrue("Process-2 is not attached properly as no threads found for it", ctx != null && ctx.length > 0);
+	}
+
+	@Test
+	public void launchGdbServerWithoutProgram_DebugNewExecutable_And_ConnectExecutable_noProgram() throws Throwable {
+		launchGdbServerWithProgram_DebugNewExecutable_And_ConnectExecutable();
+	}
+
+	@Test
+	public void launchGdbServerWithProgram_ConnectExecutable_And_DebugNewExecutable() throws Throwable {
+		assertEquals("Wrong number of containers at the start of launch", SyncUtil.getAllContainerContexts().length,
+				getInitialProcessCount());
+
+		Process process = launchApplication();
+		IMIExecutionDMContext[] ctx = attachToProcess(Long.toString(process.pid()));
+		assertTrue("Process-1 is not attached properly as no threads found for it", ctx != null && ctx.length > 0);
+
+		final IPath execPath = fGDBBackend.getProgramPath();
+		Map<String, Object> attributes = getLaunchConfiguration().getAttributes();
+
+		ctx = debugNewProcess(execPath, attributes);
+		assertTrue("Process-2 is not started properly as no threads found for it", ctx != null && ctx.length > 0);
+	}
+
+	@Test
+	public void launchGdbServerWithoutProgram_ConnectExecutable_And_DebugNewExecutable_noProgram() throws Throwable {
+		launchGdbServerWithProgram_ConnectExecutable_And_DebugNewExecutable();
+	}
+
+	private int getInitialProcessCount() {
+		return Boolean.valueOf((String) getLaunchAttribute(ITestConstants.LAUNCH_GDB_SERVER_WITHOUT_PROGRAM)) ? 0 : 1;
+	}
+
+	private Process launchApplication() throws IOException {
+		Process process = Runtime.getRuntime().exec(getBinary());
```

Comment:
Wouldn't it be better to use the CDT ProcessFactory here?

Reply:
no strong opinion.. we just need to launch a program and then attach to it.. IIUC, ProcessFactory is useful when there is some signal handling and pseudo-terminal related stuff is required..

Reply:
I just wrote a long comment on this that github just trashed.

On Windows (I assume that the tests are executed on Windows as the extension is set to ".exe" for the test application), sending 2 signals to an application will just have the process vanish from the process list without letting it do cleanup of any shared resources. When calling Process.destroy() in java, the native call is going to be a call to [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) and reading the documentation, it sounds just as bad as sending 2 signals.
To avoid having stale handles etc on the server, I suggest we use the CDT way instead that has been extensively tested over the years to handle this exact case.

For reference: [Process.destroy0 on Windows](https://github.com/openjdk/jdk17/blob/master/src/java.base/windows/native/libjava/ProcessHandleImpl_win.c#L369) and [Process.destroy0 on *nix](https://github.com/openjdk/jdk17/blob/74007890bb9a3fa3a65683a3f480e399f2b1a0b6/src/java.base/unix/native/libjava/ProcessHandleImpl_unix.c#L316)

Reply:
The tests haven't been maintained on Windows for a while, they once passed but not sure of current state. The `.exe` is explained here:

https://github.com/eclipse-cdt/cdt/blob/d2a5a57d01078a7ccf5919f2a429ba197dafec0a/dsf-gdb/org.eclipse.cdt.tests.dsf.gdb/data/launch/src/Makefile#L17

That said, we should indeed be launching with process factory, here is an example of how gdbserver is launched in the base test case:

https://github.com/eclipse-cdt/cdt/blob/d2a5a57d01078a7ccf5919f2a429ba197dafec0a/dsf-gdb/org.eclipse.cdt.tests.dsf.gdb/src/org/eclipse/cdt/tests/dsf/gdb/framework/BaseTestCase.java#L689

Reply:
> On Windows (I assume that the tests are executed on Windows as the extension is set to ".exe" for the test application)

They are run on linux only.. see `MultiProcessRemoteTest.beforeClass()` and `BaseTestCase.supportsNonStop()`.. regarding exe extension, see following

https://github.com/eclipse-cdt/cdt/blob/81581c2116f7b0a557493a2b3102bc68a4e070f9/dsf-gdb/org.eclipse.cdt.tests.dsf.gdb/data/launch/src/Makefile#L17

> sending 2 signals to an application will just have the process vanish from the process list without letting it do cleanup of any shared resources. 

I need to understand how two signals are sent, even its just the case of windows (just for my knowledge)?

Reply:
> > sending 2 signals to an application will just have the process vanish from the process list without letting it do cleanup of any shared resources.
> 
> I need to understand how two signals are sent, even its just the case of windows (just for my knowledge)?

https://github.com/eclipse-cdt/cdt/blob/7c6bd5bdcb4cf8c31050aeeb102250a27a157728/core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/Spawner.java#L334

As you can see in this method, both the `terminate()` and `kill()` methods are called and both of them sends a signal to the application.
If you want to see what happens, you can simply create a C/C++ application and have it register a signal handler and then sleep for  1 min or something. When running the application in a DOS prompt, try to hit `Ctrl + c`, wait a few sec and then hit `Ctrl + c` again and see what happens. To be more visual confirmation, you can have a printout in the signal handler to see if it's called or not.

Reply:
> have it register a signal handler

thats what I said in my first response that its useful when we are dealing with application which does special handling for signals like GDB and gdbserver does, but it was not the case for launching application here..

but thanks for the explanation.. I now understand how Spawner handles such cases..